Update app engine flexible deployment UX

etanshaul opened the pull request:
Fixes #1477 

Updates the remainder of the deployment UX.

Screenshots coming..

Still missing test coverage refresh

This PR updates the flex config portion of the deployment UI (the portion with the app.yaml and dockerfile config) to match the refresh that was done in #1476.

Also, as discussed offline, I got rid of the previously available app.yaml override and Dockerfile override features in favor of clarity and simplicity.

etanshaul commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -78,7 +78,7 @@ appengine.flex.configtype.custom.label=Custom
 appengine.flex.dockerfile.location.label=Dockerfile path\:
 appengine.flex.version.placeholder.text=auto generated
-appengine.flex.facet.config.title=App Engine flexible configuration
+appengine.flex.facet.config.title=App Engine Flexible Configuration
```
This appears as a panel title, so I think its ok to have "Flexible" capitalized. Feel free to disagree.

etanshaul commented:
No modules with flex support:
![image](https://user-images.githubusercontent.com/1735744/26897885-0a8242c8-4b98-11e7-9c6c-22c94603d34c.png)

app.yaml with runtime: java (the edit button links to the module settings):
![image](https://user-images.githubusercontent.com/1735744/26899308-bd1b56a0-4b9c-11e7-90bd-f3ab5113463f.png)

app.yaml with runtime: custom (the Dockerfile path links to the module settings):
![image](https://user-images.githubusercontent.com/1735744/26899381-f0395488-4b9c-11e7-83b5-a03aa8b648a3.png)

Showing multiple app.yamls (from different modules) - paths are displayed relative to the project root:
![image](https://user-images.githubusercontent.com/1735744/26899402-01857afa-4b9d-11e7-891e-3c97d4257062.png)

etanshaul commented:
See the above screenshots for an overview of what things look like.

Things **not** fixed here (will file separate issues):
- Move error messages closer to issue. Notice how now they still show up at the bottom:
![image](https://user-images.githubusercontent.com/1735744/26899527-667fc910-4b9d-11e7-86f2-4a82dbaaa3c9.png)

- Don't block saving the dialog on errors. Currently its still preventing the user from saving until errors are resolved.

nkibler commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java:
```diff
@@ -147,116 +129,92 @@ protected void textChanged(DocumentEvent event) {
+    appYamlCombobox.setRenderer(new ListCellRendererWrapper<AppEngineFlexibleFacet>() {
       @Override
-      public void customize(JList list, Module value, int index, boolean selected,
+      public void customize(JList list, AppEngineFlexibleFacet facet, int index, boolean selected,
```
nit: Use the Google Java formatter to fix this, but the params should either be altogether on the next line (indented by 4 spaces), or if they don't fit like that either, all on separate lines indented by 4 spaces.

nkibler commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditor.java:
```diff
@@ -147,116 +129,92 @@ protected void textChanged(DocumentEvent event) {
+        runtimePanel.setLabelText(flexibleRuntime.map(FlexibleRuntime::toString).orElse(""));
+        runtimePanel.setVisible(true);
+        runtimePanel.setExplanationLabelVisibility(
+            !flexibleRuntime.map(FlexibleRuntime::isCustom).orElse(false));
```
Why not use `isCustomRuntime()` here since this logic is already separated into its own method?

etanshaul replied:
I had a mental todo to look into refactoring this if possible.

But the reason is that if you look a few lines up, I already extract the flexRuntime. I didn't want to call isCustomRuntime which would just duplicate this extraction (which requires IO - parsing the app.yaml file).

nkibler commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/AppEngineFlexibleRuntimePanel.java:
```diff
@@ -0,0 +1,62 @@
+/**
+ * UI panel containing App Engine flexible runtime information.
+ */
+public class AppEngineFlexibleRuntimePanel {
```
final?

nkibler commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/AppEngineFlexibleRuntimePanel.java:
```diff
@@ -0,0 +1,62 @@
+  private JLabel runtimeExplanationLabel;
+
+  public AppEngineFlexibleRuntimePanel() {
+    runtimeExplanationLabel.setFont(
```
How is this guaranteed to not be null here?

etanshaul replied:
its part of the magic of IJ UX form binding. This class is bound to a GUI form of the same name (with a .form) extension. The platform is responsible for initializing the class properly.

patflynn replied:
It's pretty weird, especially when you don't see a marker interface or anything to indicate magic.

nkibler commented on google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/appengine/cloud/flexible/AppEngineFlexibleDeploymentEditorTest.java:
```diff
@@ -242,117 +214,6 @@ public void testValidateConfiguration_invalidCloudSdk() {
     }
   }
 
-  public void testValidateConfiguration_blankYaml() {
```
Why remove these tests? Are they not valid anymore?

etanshaul replied:
They are not - all these tests were removed as part of the removal off the app.yaml / Dockerfile override feature mentioned in the first comment of the PR.

I still have a todo to fix some tests and add more coverage for the changes.

etanshaul commented:
I just noticed that with these changes, the flex cost label location didn't make sense anymore. I moved it to be underneath the environment label (and added an icon to it). Of course this works as expected for flex compat too.

![image](https://user-images.githubusercontent.com/1735744/26902949-2d59a906-4ba9-11e7-964c-c3d046b3f78a.png)

_Note: I think prior to merging this we can give it a quick look over to see if we can optimize the spacing / margins in the layout everywhere_

patflynn commented:
@etanshaul This is looking pretty slick! Starting review noww

patflynn commented on google-cloud-tools-plugin/resources/messages/CloudToolsBundle.properties:
```diff
@@ -139,20 +140,20 @@ appengine.stop.modules.version.success.dialog.message=Application has been succe
 appengine.deployment.error.invalid.cloudsdk=Invalid Cloud SDK directory path.
 appengine.deployment.deployallconfigs.checkbox.label=Update dos, dispatch, cron, queues, and datastore indexes
-appengine.deployment.flexible.nogaesupport=No modules with Google App Engine support found. Please provide your configuration files
+appengine.deployment.flexible.nogaesupport=No modules with Google App Engine flexible support found.
```
Is this caused by the lack of the facet? Can we be specific about that? e.g. No modules with the GAE Flex facet found..?

etanshaul replied:
yea its caused by no modules / no modules with flex facet.

I hesitated to mention the facet since the way users are typically expected to add facets is via framework support.

Maybe I should mention framework support more explicitly:
No modules with Google App Engine **framework** support found.

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/AppEngineFlexibleFacet.java:
```diff
@@ -39,4 +41,16 @@ public AppEngineFlexibleFacet(@NotNull FacetType facetType,
+   * isn't one.
+   */
+  @Nullable
+  public static AppEngineFlexibleFacet getAppEngineFacetByModule(@Nullable Module module) {
```
Why accept a null module?

patflynn replied:
why accept null module params?

etanshaul replied:
good point. updated

patflynn commented on google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/AppEngineFlexibleFacet.java:
```diff
@@ -39,4 +41,16 @@ public AppEngineFlexibleFacet(@NotNull FacetType facetType,
   }
+
+  /**
+   * Returns the {@link AppEngineFlexibleFacet} present in the supplied module, or null if there
```
I think literals like null should be {@code null}

patflynn commented:
a couple of UI thoughts:
- the italics message about setting a custom runtime looks kind of weird as a child of runtime.  Maybe we could pull it up to a footnote of the panel?  
- Do all run config validation errors always have 'invalid data' prefixes/

etanshaul commented:
`the italics message about setting a custom runtime looks kind of weird as a child of runtime. Maybe we could pull it up to a footnote of the panel?`
My thought was that this message is directly applicable to runtimes so I had it under the runtime section. I don't feel too strongly about it though.

`Do all run config validation errors always have 'invalid data' prefixes`
Yea this is part of what the framework does when throwing ConfigurationExceptions for validation. This will be removed in a future PR where we no longer block on form errors.

etanshaul commented:
@patflynn tests have been fixed and updated, and comments addressed. only potential todo is moving the location of the custom runtime / Dockerfile message.

etanshaul commented:
@patflynn 
here's what the message looks like un-nested:
![image](https://user-images.githubusercontent.com/1735744/26946225-9b490a70-4c5c-11e7-9614-280e1fc0af62.png)

wdyt? maybe remove the italics too..

patflynn commented:
hmm, what do you think about removing the message completely?

On Thu, Jun 8, 2017 at 3:10 PM, Etan Shaul <notifications@github.com> wrote:

> @patflynn <https://github.com/patflynn>
> here's what the message looks like un-nested:
> [image: image]
> <https://user-images.githubusercontent.com/1735744/26946225-9b490a70-4c5c-11e7-9614-280e1fc0af62.png>
>
> wdyt? maybe remove the italics too..
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1493#issuecomment-307199172>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5HRPT86oo-mekieiASLxMuHUolhfbks5sCEcygaJpZM4NzK0D>
> .
>

etanshaul commented:
done e1ece90

![image](https://user-images.githubusercontent.com/1735744/26946914-d79792e2-4c5e-11e7-9bb2-0b4c9df3d635.png)

etanshaul commented:
found a bug: saving the configuration doesn't save the selected app.yaml in the dropdown. The path is saved, but need to restore the correct selection in the combobox. Fix coming.

patflynn commented:
the path to the app.yaml is saved?  shouldn't it be read from the module
facet at execution time?

On Thu, Jun 8, 2017 at 3:28 PM, Etan Shaul <notifications@github.com> wrote:

> found a bug: saving the configuration doesn't save the selected app.yaml
> in the dropdown. The path is saved, but need to restore the correct
> selection in the combobox. Fix coming.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-intellij/pull/1493#issuecomment-307203595>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHf5Hc5Iwb0TTDM4016arZL-he3oXldWks5sCEtWgaJpZM4NzK0D>
> .
>

etanshaul commented:
hmm currently what's persisted is the app.yaml path itself. this is sort of a legacy thing brought over from when we allowed app.yaml overrides.

Its not technically wrong, but instead we could persist the selected facet instead (the facet corresponding to the selected app.yaml in the dropdown)

etanshaul commented:
on second thought, changing what's persisted would affect user migration. its safer to keep storing the app.yaml path as is.

etanshaul commented:
ahh. I see your point. We have the potential now for the app.yaml (and Dockerfile for that matter) to be stale. Ill look into this now.

etanshaul commented:
some things to note about the above:

Once fixed, and released, users upgrading that have a prior custom flex deployment will have to re-enter the app.yaml (and Dockerfile if applicable). This is a consequence of moving these to the facet.

Attempting to do some auto migration like checking to see if there is a stored app.yaml / Dockerfile in the configuration and then moving it to the facet wouldn't be robust as there isn't a one-to-one mapping between deployment config and facet config.
